### PR TITLE
Add a GUC to discard redistribute hashjoin for Orca

### PR DESCRIPTION
**Pull request by gpopt**
Add a GUC `optimizer_discard_redistribute_hashjoin` to control redistribute
hashjoin.

If any one of HashJoin's children operators is RedistributeMotion, this plan
alternative will be given MAX cost, hence it will not be picked by optimizer.
Default value is `OFF`.

### CODE REVIEW

**Comment by chrishajas** [comment]
A couple things:
1) This isn't disabling as we've done in the past, rather it's preferring not to since we're using the cost model rather than disabling the alternative. Is that what we want here? Should it be named differently?
2) Using costing with max floats can have some consequences if those are the only alternatives produced. There might not be any distinction between plan costs at that point, and we may end up choosing a higher costed plan.
3) I think we're adding this because we need a knob and don't have plan hints?

**Comment by gpopt** [comment]
Good questions!
1. You are right. It is preventing it being selected. Giving it a MAX cost will eventually disable it, even the alternative may still be generated, but will not be selected. Legacy planner, on the other hand, `enable_nestloop` is not actually disabling it, just giving it some penalty (1000x). I can try to disable it in the property enforcement instead of costing the operator MAX, so that the alternative won't be generated at all. WDYT?
2. HashJoin with Redistribute won't be the only option, unless we disable broadcast and gather motion at the same time. If that's the case, blame the user, who is burning himself/herself. :)
3. Correct, I hope we had a hint to control for a single join, instead of affecting all the HashJoin nodes in a single plan.

**Comment by gpopt** [comment]
Hmm, it maybe difficult to disable it in the property enforcement framework, because we can only get the operator and the required physical property from parent, not what type of the parent operator is. We need to know it is HashJoin and the distribution property needs to be enforced.

**Comment by chrishajas** [comment]
> Good questions!
> 
> 1. You are right. It is preventing it being selected. Giving it a MAX cost will eventually disable it, even the alternative may still be generated, but will not be selected. Legacy planner, on the other hand, `enable_nestloop` is not actually disabling it, just giving it some penalty (1000x). I can try to disable it in the property enforcement instead of costing the operator MAX, so that the alternative won't be generated at all. WDYT?
> 2. HashJoin with Redistribute won't be the only option, unless we disable broadcast and gather motion at the same time. If that's the case, blame the user, who is burning himself/herself. :)
> 3. Correct, I hope we had a hint to control for a single join, instead of affecting all the HashJoin nodes in a single plan.

I think we should consistent with how we're disabling xforms (either cost based vs not generating the alternative) for now, otherwise we'll end up confusing ourselves and users at some point. It would be nice to have everything as cost based, but I'd prefer not having a mix (at least with the same name)

> Hmm, it maybe difficult to disable it in the property enforcement framework, because we can only get the operator and the required physical property from parent, not what type of the parent operator is. We need to know it is HashJoin and the distribution property needs to be enforced.

We do something sorta similar I think with `EopttraceEnableRedistributeBroadcastHashJoin`, though we may need to add logic in a couple places.

**Comment by gpopt** [comment]
@chrishajas I checked the code, `EopttraceEnableRedistributeBroadcastHashJoin` doesn't actually disable redistribute broadcast hashjoin, it just controls whether we want to pass down the required hash distribution property from parent operator to its outer child. Even if we turned it off, Orca may still generate hashjoin with outer redistribute and inner broadcast.

In order to implement it in the optimizer framework, we need to request a hash distribute request that is not enforceable, which Orca doesn't support yet. We only had done it in https://github.com/greenplum-db/gpdb/pull/13833. I don't think it is worth the effort for this case.

**Comment by gpopt** [comment]
But if you think the naming is not appropriate, we can definitely rename the GUC to a more meaningful name.

**Comment by chrishajas** [comment]
> @chrishajas I checked the code, `EopttraceEnableRedistributeBroadcastHashJoin` doesn't actually disable redistribute broadcast hashjoin, it just controls whether we want to pass down the required hash distribution property from parent operator to its outer child. Even if we turned it off, Orca may still generate hashjoin with outer redistribute and inner broadcast.
> 
> In order to implement it in the optimizer framework, we need to request a hash distribute request that is not enforceable, which Orca doesn't support yet. We only had done it in #13833. I don't think it is worth the effort for this case.

I see... I'm ok with adding it since we don't have an alternative currently. But we really do need to get plan hints or something so that we don't have such specific gucs like this.

As for the costing implementation/naming, we may want to get PM opinion. I'm most concerned about developer and user confusion if we have multiple of these. Some will be cost based and others will be disallowing the alternative entirely.

**Comment by gpopt** [comment]
As we discussed offline, we will rename the GUC to `optimizer_discard_redistribute_hashjoin`, which better describes what this GUC is doing.

File: src/backend/gporca/libnaucrates/include/naucrates/traceflags/traceflags.h [hunk 1]
```diff
@@ -215,6 +215,9 @@ enum EOptTraceFlag
 
 	EopttraceForceComprehensiveJoinImplementation = 103041,
 
+	// Discard HashJoin with RedistributeMotion nodes
+	EopttraceDiscardRedistributeHashJoin = 103044,
```
Comment (dgkimura) [inline src/backend/gporca/libnaucrates/include/naucrates/traceflags/traceflags.h:219]: Trace flag numbers seem to skip 103042 and 103043.  I think due to cherry-pick from 6X to main.

**Review by dgkimura — APPROVED** [review]
Looks good to me.  I strongly agree with Chris that the ultimate solution is to develop plan hints or some way to force query shapes and then deprecate a lot of these query tuning GUCs.

File: src/backend/gporca/libnaucrates/include/naucrates/traceflags/traceflags.h (hunk 1 above)
Comment (gpopt) [reply src/backend/gporca/libnaucrates/include/naucrates/traceflags/traceflags.h:219]: Yes, you have a sharp eye. I think it is better to keep them consistent.

**Comment by gpopt** [comment]
> I strongly agree with Chris that the ultimate solution is to develop plan hints or some way to force query shapes and then deprecate a lot of these query tuning GUCs.

Agreed, now we are allowed to work on the plan hint (Orca only) next year!

**Review by khannaekta — APPROVED** [review]
LGTM!

**Comment by gpopt** [comment]
Merged into main branch, thanks all for the review!

**Comment by gpopt** [comment]
@lisakowen Documentation draft:
The optimizer GUC `optimizer_discard_redistribute_hashjoin` controls whether to discard plan alternatives where the HashJoin has a Redistribute Motion child.

The default value is `OFF`.

**Pull request merged**